[RFC] api: nvim_get_mode()

justinmk opened the pull request:
Until now, asynchronous API functions never flush the input queue. `nvim_get_mode()` is different: 

- if RPCs are known to be blocked, it responds immediately (without flushing the input/event queue)
- else it flushes the input/event queue in order to determine whether pending input would cause RPCs to block, then responds with the result

Internally this is handled as a special case, but _semantically_ nothing has changed: API users never know when input flushes, so this internal special-case doesn't violate that. As far as API users are concerned, `nvim_get_mode()` is just another asynchronous function.

In all cases `nvim_get_mode()` never blocks for more than the time it takes to flush the input/event queue (~µs).

Note that this PR doesn't address #6166, and `nvim_get_mode()` will provoke #6166 if e.g. `d` is operator-pending. That will be the next thing I look at.

Comments from anyone who groks the event loop (@bfredl @tarruda @oni-link) would be very welcome.

Closes #6159
cc @lunixbochs

justinmk commented on src/nvim/msgpack_rpc/channel.c:
```diff
@@ -430,6 +432,13 @@ static void handle_request(Channel *channel, msgpack_object *request)
+    char buf[256] = { 0 };
+    memcpy(buf, method->via.bin.ptr, MIN(255, method->via.bin.size));
+    bool is_get_mode = strcmp("nvim_get_mode", buf) == 0;
+    handler.async = is_get_mode && input_blocking();
```
special case `nvim_get_mode`: If we know RPC will be blocked, it can (and must) be asynchronous.

justinmk commented on src/nvim/os/input.c:
```diff
@@ -323,13 +325,26 @@ static unsigned int handle_mouse_event(char **ptr, uint8_t *buf,
+      && !(events_enabled || input_ready() || input_eof)
+      ) {
+    blocking = true;
+    multiqueue_process_events(main_loop.events);
```
We found out that the pending input will provoke a blocking wait, so here we process queued events (which may include the synchronous `nvim_get_mode`) once before blocking.

tarruda commented:
It's been quite some time since I looked through the code, so I may be missing something and if so, please let me know.

I don't think I understand this PR, why do you need a function marked with `FUNC_API_ASYNC` to be `sync` in some cases? If the goal is simply allow embedder to determine the current editor mode (and I assume the reason for that is to know if a certain request would block) why do you need to flush the input buffer?

tarruda commented on src/nvim/rbuffer.c:
```diff
@@ -34,17 +34,20 @@ void rbuffer_free(RBuffer *buf)
 
-size_t rbuffer_space(RBuffer *buf) FUNC_ATTR_NONNULL_ALL
+size_t rbuffer_space(RBuffer *buf)
+  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_PURE
```
Are you sure it is safe to mark these functions as pure? Considering `RBuffer` is used by both the main thread and the ui thread, the compiler may not be able to when `RBuffer` is changed by another thread.

justinmk replied:
Thanks for mentioning that. I didn't know that pure is not allowed to be used if the parameter is used by multiple threads, that makes it extremely difficult to ever know if a function can safely be marked pure. @ZyX-I can you confirm?

tarruda replied:
Maybe the compiler will consider the memory as changed if we use locks to access `RBuffer`(which we do, if I remember correctly), but then this attribute won't be of much help. In any case, since the computation is very trivial, it is better to be safe than sorry(trying to debug it in a far future, where we forgot about this)

justinmk replied:
Agreed, I'll remove this change. Just curious for future reference, though.

justinmk commented:
> If the goal is simply allow embedder to determine the current editor mode (and I assume the reason for that is to know if a certain request would block) why do you need to flush the input buffer?

That is the goal, in particular the RPC client needs to know if the next synchronous RPC will block. There could be input in the queue which wasn't flushed yet, which will cause the next synchronous to RPC block. For example:

```
nvim_input('g')
nvim_get_mode()
nvim_command('foo') // this will block
```

In the above sequence, if `nvim_get_mode()` does not flush the input queue, it will incorrectly return `blocking=false`. Based on that result, the client thinks it can safely call `nvim_command()`, which _will_ flush the queue, and now the client is blocked. 

For single-threaded clients like [actualvim](https://github.com/lunixbochs/actualvim), this is a problem because user input is sent by the client. If the client is blocked then it's stuck.

tarruda commented on src/nvim/os/input.c:
```diff
@@ -323,13 +325,25 @@ static unsigned int handle_mouse_event(char **ptr, uint8_t *buf,
   }
 
+  if ((ms == - 1 || ms > 0)
+      && !(events_enabled || input_ready() || input_eof)
```
Because it has been quite some time since I've maintained this code, I'm not sure what are the implications of adding this `multiqueue_process_events()` call, but consider that `input_poll` is called no matter what state nvim is in, so it may not be safe to process events.

In any case, this condition is the same as `(ms == - 1 || ms > 0) && !events_enabled && !input_ready() && !input_eof`, but `input_ready()` cannot be true for this function to be called(See `input_poll` below), so it could be simplified to `(ms == - 1 || ms > 0) && !events_enabled && !input_eof`

justinmk replied:
I was afraid of that. The only other option I can think of is to manage the `nvim_get_mode` requests in a child queue, and then process only those calls here. E.g.

    channel_process_safe_events()

or something like that. And that would iterate through `nvim_get_mode` events. 

I thought maybe `main_loop.fast_events` could be used for this purpose somehow, but didn't have success.

tarruda commented:
> There could be input in the queue which wasn't flushed yet, which will cause the next synchronous to RPC block.

OK I understand the issue now. Even so, I'm unsure if this is the best approach.

When I started the pushdown automaton refactor(@8d93621c6368fb6e3e3f7138bff50e08585f347b), the goal was to add explicit support for `K_EVENT` in more states, in the hopes we can eventually remove `FUNC_ATTR_ASYNC`. Maybe we can try to follow this direction instead?

For now I suggest to investigate the amount of work required to add support for `K_EVENT` to the various operator pending modes(assuming operator pending is the main source of problems here), and if it is a lot of work, try to create a plan for refactoring. I actually had a lot of ideas on how to remodel the code to make this easier, but I need to refresh memory first before elaborating further.

justinmk commented:
I didn't forget about the K_EVENT refactoring intentions, but that's a high-risk task. Adding a new API function is much lower risk, and also forward-compatible with whatever future refactoring may occur.

We also need to un-block https://github.com/lunixbochs/actualvim (and who-knows-what else: this has been a known issue for ~2 years).

justinmk commented:
Any comments on https://github.com/neovim/neovim/pull/6247#discussion_r105387285 ? Could `fast_events` be leveraged somehow?

tarruda commented:
> I didn't forget about the K_EVENT refactoring intentions, but that's a high-risk task. Adding a new API function is much lower risk, and also forward-compatible with whatever future refactoring may occur.

It may not be as high-risk as you think, I actually had written down a plan on how to incrementally move vim implicit modes into the `state_enter` loop(the plan is written in the commit message for(@8d93621). In fact, I think it is safer to create an explicit states for some of the operator-pending modes than to process events in API calls marked as async.

> Any comments on #6247 (comment) ? Could fast_events be leveraged somehow?

I'm currently doing some reading on the code and on the related issues to refresh my memory and will come back with a response later today.

lunixbochs commented:
ActualVim will now detect and use this function if available. It works great for me, and is reliably faster than my heuristic method.

https://github.com/lunixbochs/actualvim/commit/8a21e683723bc3b65f9b753cbf2e92edee6b856a#diff-b00fc497b0226e8afc26fccf90e2c4b0R300

tarruda commented:
A small update: I'm currently working on a PR that aims to fix #6166/#6159 by doing a ambitious refactoring of normal mode, but still need some days before have something to show. This is something I had planned in 2015 but ended up never finishing, and reading the code on friday/saturday helped me refresh my memory.

It is not the simple refactoring I [commented about](https://github.com/neovim/neovim/pull/6247#issuecomment-285668655), but I strongly believe it is the best solution for all problems related to async integration with nvim's main loop(it might even allow removing the distinction between immediate/deferred RPC calls in the short term). It is also something that will greatly help in cleaning up some other very old pieces of code.

Right now it is hard to estimate how many more days I need to finish this, but I will have another update by the end of the week.

justinmk committed:
api: nvim_get_mode()

Asynchronous API functions are served immediately, which means pending
input could change the state of Nvim shortly after an async API function
result is returned.

nvim_get_mode() is different:
  - If RPCs are known to be blocked, it responds immediately (without
    flushing the input/event queue)
  - else it is handled just-in-time before waiting for input, after
    pending input was processed. This makes the result more reliable
    (but not perfect).

Internally this is handled as a special case, but _semantically_ nothing
has changed: API users never know when input flushes, so this internal
special-case doesn't violate that. As far as API users are concerned,
nvim_get_mode() is just another asynchronous API function.

In all cases nvim_get_mode() never blocks for more than the time it
takes to flush the input/event queue (~µs).

Note: This doesn't address #6166; nvim_get_mode() will provoke #6166 if
e.g. `d` is operator-pending.

Closes #6159

justinmk committed:
input.c: Process only safe events before blocking.

Introduce multiqueue_process_priority() to process only events at or
above a certain priority.

justinmk commented:
@tarruda That is great news. I pushed an update to this PR that abuses `Event.priority` so that `input_poll()` only processes async events. That should be safe, right?

If that is sound I'm thinking we should merge this and can remove the ugly stuff once the pretty stuff is ready.

justinmk commented on src/nvim/event/multiqueue.c:
```diff
@@ -141,13 +143,40 @@ void multiqueue_process_events(MultiQueue *this)
+      multiqueue_remove(this);
+    } else {
+      // not processed
+      cur = cur->next->next;
```
is this correct?

oni-link commented on src/nvim/api/vim.c:
```diff
@@ -664,6 +666,20 @@ Dictionary nvim_get_color_map(void)
+  char* buf = get_mode();
+  bool blocked = input_blocking();
+
+  ADD(rv, STRING_OBJ(cstr_to_string(buf)));
```
`cstr_as_string()` should be used to prevent leaking the `buf` string.

tarruda commented:
> I pushed an update to this PR that abuses Event.priority so that input_poll() only processes async events. That should be safe, right?

It is difficult to tell with certainty. Apparently it is safe, but I think it is irrelevant.

I suspect solution implemented here doesn't work. If I understood correctly, you are trying to ensure `nvim_get_mode` is only handled after nvim main loop has fully processed all pending input, but the code you added here doesn't seem to guarantee that. Let's consider the example you gave before:

```
nvim_input('g')
nvim_get_mode()      // "g" hasn't been processed yet
nvim_command('foo')  // this will block
```

Imagine the following chain of events:

- Nvim calls os_inchar to wait for input or events
- RPC for `nvim_input` is received, which is handled immediately by posting data to the input buffer

What happens next depends on OS races that are out of our control. The two possible scenarios are: 

- The OS buffered `nvim_get_mode` before libuv poll returns to process `nvim_input`. In this case, `nvim_get_mode` will be processed in the same loop as `nvim_input`.
- The OS buffered `nvim_get_mode` after libuv poll returns to process `nvim_input`. In this case, `nvim_get_mode` will be processed in the next `os_inchar` call.

The only way you will get the result you expect(`nvim_get_mode` returns the correct mode, which is pending 'g' command), is in the second scenario, but calling `multiqueue_process_event` is irrelevant since the queue will be empty in any of the two situations. Not only the processing of `nvim_input` happened before `nvim_get_mode`, but even if it was after, processing `nvim_input` would simply put data to input buffer, not force the main loop to process it.

I could be missing something, of course(again, it has been a long time :)), but my theory can be verified by making sure that the requests for `nvim_input` and `nvim_get_mode` are written to nvim pipe in a single batch(assuming the OS pipe buffer is large enough, which is probably true).

Note: Right now `Event.priority` is not used, and if I remember correctly it was created it before the parent/child queue concept, which gives a better way to selectively process events. The reason `Event.priority` still exists is because forgot to remove it, so it should be removed it in a future PR.

justinmk commented:
> The only way you will get the result you expect(nvim_get_mode returns the correct mode, which is pending 'g' command), is in the second scenario, but calling `multiqueue_process_event` is irrelevant since the queue will be empty in any of the two situations.

It can't be irrelevant, because commenting out that line in input.c causes the tests to hang (at the `nvim_get_mode` RPC call).

justinmk commented:
>  my theory can be verified by making sure that the requests for nvim_input and nvim_get_mode are written to nvim pipe in a single batch(assuming the OS pipe buffer is large enough, which is probably true).

Confirmed. If nvim_get_mode is in the same batch, it's useless. Demonstrated via `nvim_call_atomic`: https://github.com/neovim/neovim/pull/6247/commits/36a7d3d06879dfff434f0c3859d6778e854fc033#diff-5bf434f0a1fee647b4b121473e3410bfR280

tarruda commented:
> It can't be irrelevant, because commenting out that line in input.c causes the tests to hang (at the nvim_get_mode RPC call--i.e. #6166).

Maybe because of the logic you added that makes `nvim_get_mode` not be an async call in some situations?

justinmk commented:
> Maybe because of the logic you added that makes nvim_get_mode not be an async call in some situations?

The purpose of that logic is to flush the input queue. If `input_blocking()` is false and there is pending input, _asynchronous_ `nvim_get_mode` is useless.

tarruda commented:
> Which means in that scenario (which hasn't been observed in the tests, yet) the worst that can happen is what already happens 100% of the time on current master.

True, but why add an API whose purpose is to test if the next call would block, but that has a possibility of returning wrong information? If the user can't rely on the return value 100% of times, he will probably not use it.

Even if you still want to add an API that allows the user to sometimes know if the next call would block, I think there's a much simpler way: You can implement a function marked as ASYNC that returns the state of the `events_enabled` variable.

justinmk commented:
@tarruda Does https://github.com/neovim/neovim/pull/6247/commits/36510c9b1f66e09bef439650a9b16346654af416 look right? Doesn't seem to work.

tarruda commented:
> @tarruda Does 303897b look right? Doesn't seem to work.

I will have a look

justinmk committed:
api/nvim_get_mode: Use child-queue instead of "priority".

justinmk committed:
event: Remove "priority" concept.

It was replaced by the "child queue" concept (MultiQueue).

justinmk commented:
@lunixbochs I changed the result of nvim_get_mode to be a dictionary:

    { "mode": String, "blocking: Boolean }

Going to merge this today.
